An initial support for FreeBSD

robertek opened the pull request:
This is an initial support for the FreeBSD.

The source can be built on FreeBSD and Linux. Tested both debug and release build.

On FreeBSD it is working with some exceptions.
1) I have skipped the lua support for now. So it can be built only without it. I will add it as next PR, it was not a priority.
2) The timers are not supported for now, since the missing timerfd(2) syscall in FreeBSD <=13. It was added in the FreeBSD 14, which isn't released yet.

I understand if the timerfd patch would not be accepted, it is little bit a nasty hack. I may add it as patch to the ports tree when I will submit the myMPD port. I may also implement this patch somehow else, but since it is only a temporary fix, I haven't seen a point in much more complex fixing.

I will not submit the port tree Makefile in this repo, I don't see a point in it. But if you want to have it here as a reference, I will add it. I have also created a rc.d script, which is not fully tuned yet. I may submit it also if you want, otherwise it will be submitted in the FreeBSD's port tree.

jcorporation commented:
myMPD on FreeBSD? This is great!

I have not developed myMPD with cross plattform compatibility in mind

Master is the myMPD stable branch, we should merge your patches in a separate branch. I created a freebsd branch, please rebase your pull request and leave the timerfd patch out.

Your timerfd patch simply deactivates the timer backend, this leads to some severe problems. Besides of the user point of view, myMPD uses timers internally.

Please add the port tree Makefile and the rc.d script for reference.

We can merge the freebsd and the v13.0.0 branch into devel as soon as the v13.0.0 branch is feature complete and roughly tested. If the release schedule of FreeBSD v14 is valid (23 October 2023), it should be released before myMPD v13.0.0 is released. With this in mind, we can simply ignore the timerfd issue.

robertek commented:
Ok I will merge that to the branch and also attach the building files and rc script.

I know the timers are used even outside of the "timers" user function. As I'm running the patched version for few days in a FreeBSD jail I see some probable outcomes of this (for example never finished album index) and also the failed creation of the timers in the log. But anyway, even with this issues, I still consider them as minor, otherwise it works pretty well.

Even when the FreeBSD 14 is released the ports tree is version agnostic and should work on any of the version. So I will probably need to attach the patch to the tree. The only difference would be that on 14 it will work fully as it is supposed to, the 13 will be here for some time and myMPD would work just acceptably (form my point of view).

jcorporation commented:
I do not like myMPD builds that works just acceptable. I think we should go the way to fix myMPD to work without problems on older FreeBSD versions.

It is acceptable to disable the timer feature for the web interface, but at minimum the album indexing must work.

There are only three internal functions that uses the timers:

- TIMER_ID_COVERCACHE_CROP: unimportant
- TIMER_ID_SMARTPLS_UPDATE: unimportant
- TIMER_ID_CACHES_CREATE: important, but it could be fixed relatively easy.

Following code in `/src/mpd_client/idle.c`, Line 493

```
return mympd_api_timer_replace(&mympd_state->timer_list, timeout, TIMER_ONE_SHOT_REMOVE,
            timer_handler_by_id, TIMER_ID_CACHES_CREATE, NULL);
```

must be replaced with:

```
struct t_work_request *request = create_request(-1, 0, MYMPD_API_CACHES_CREATE, NULL, MPD_PARTITION_DEFAULT);
request->data = sdscat(request->data, "\"force\":false}}"); //only update if database has changed
mympd_queue_push(mympd_api_queue, request, 0);
```

This drops the debouncing of fast recurring mpd database update events, but this is ok, if we do this only for older FreeBSD versions.

I can merge the current version of your pull request, but before this can be merged into master, we must address  the following issues:

- Fix album index creation
- Disable timers in the user interface
- Add documentation (installation, compilation, missing features on older FreeBSD functions)

jcorporation reviewed (commented):
We should add a central function to set the threadname.

jcorporation commented:
I merge this pull request now. Let me refactor some internal apis, for better support of album cache creation without timers. I will also add the disabling of the ui elements

robertek commented:
Thanks,

for the update I have added the FreeBSD build files and the rc script.

And I can confirm that the change you suggested for TIMER_ID_CACHES_CREATE works. I have added it to the commit with timerfd change. 
The albums are now correctly listed.

I have a couple of TODOs:

1) add the documentation for FreeBSD build and some specific setup notes
2) test and fix the LUA support on FreeBSD
3) create a header file and probably a macro with linux threadname to centralize that.
4) test if the __FreeBSD__ ifdefs are really needed in the socket.h headers.
